tests/common/distributed.py

```diff
@@ -0,0 +1,227 @@
+from _pytest.outcomes import Skipped
+from torch.multiprocessing import Process
+
+logger = logging.getLogger(__name__)
+
+# Worker timeout after the first worker has completed.
+WORKER_TIMEOUT_IN_SECS = 120
+MASTER_ADDRESS = '127.0.0.1'
+
+
+class DistributedTest:
+    """A simulation of distributed behavior on a single machine."""
+    is_dist_test = True
+    # Number of ranks
+    world_size = 1
+    backend = 'gloo'
+    current_test = None
+    test_kwargs = None
+    port = None
+
+    def _run_test(self, request: FixtureRequest) -> None:
+        """Run the current test.
+
+        Args:
+            request (FixtureRequest): A request object with a test context
+
+        Raises:
+            ValueError: `world_size` should be of type `int`
+        """
+        # Override the marker if any
+        self._set_custom_markers(request)
+
+        # Get a free TCP Port to listen on
+        self.port = self.get_free_tcp_port()
+
+        self.current_test = self._get_current_test_func(request)
+        self.test_kwargs = self._get_test_kwargs(request)
+        if not isinstance(self.world_size, int):
+            raise ValueError('`world_size` should be of type `int`')
+        self._launch_processes(self.world_size)
+        time.sleep(0.5)
+
+    def _get_current_test_func(self, request: FixtureRequest) -> Any:
+        """Get the current test function.
+
+        Args:
+            request (FixtureRequest): A request object with a test context
+
+        Returns:
+            A test function
+        """
+        # Get the test method
+        func_name = request.function.__name__
+        return getattr(self, func_name)
+
+    def _get_test_kwargs(self, request: FixtureRequest) -> Dict[str, Any]:
+        """Get the test arguments.
+
+        Args:
+            request (FixtureRequest): A request object with a test context
+
+        Returns:
+            Dict[str, Any]: A dictionary of arguments
+        """
+        # Get fixture / parametrize kwargs from pytest request object
+        test_kwargs = {}
+        params = inspect.getfullargspec(self.current_test).args
+        params.remove('self')
+        for p in params:
+            test_kwargs[p] = request.getfixturevalue(p)
+        return test_kwargs
+
+    def _set_custom_markers(self, request: FixtureRequest) -> None:
+        """Get the custom markers and overrides it.
+
+        Args:
+            request (FixtureRequest): A request object with a test context
+        """
+        # Fetch custom pytest marker `world_size`
+        # and override with a class parameter
+        for mark in getattr(request.function, 'pytestmark', []):
+            if mark.name == 'world_size':
+                self.world_size = mark.args[0]
+
+    def get_free_tcp_port(self) -> int:
+        """Get a free socket port to listen on."""
+        # from https://www.programcreek.com/python/?CodeExample=get+free+port
+        tcp = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
```

Comment:
perhaps just socket.socket() as those args are the defaults? url mention not necessary :)

Reply:
Changed in https://github.com/mosaicml/streaming/pull/57/commits/8741970777041113ee42870ed439f33131211835.